src/ElmFormat/Render/Box.hs

```diff
@@ -2121,6 +2065,87 @@ formatComment comment =
             mustBreak $ row [ punc "{--", literal c, punc "-}" ]
 
 
+formatCaseExpression ::
+    ElmVersion
+    -> ImportInfo
+    -> (AST.Commented AST.Expression.Expr, AST.Multiline)
+    -> ([(AST.Commented AST.Pattern.Pattern, (AST.Comments, AST.Expression.Expr))], AST.Multiline)
+    -> Box
+formatCaseExpression elmVersion importInfo (subject, multilineSubject) (branches, multilineFirstBranch) =
+    let
+        opening =
+          case
+            ( multilineSubject
+            , formatCommentedExpression elmVersion importInfo SyntaxSeparated subject
+            )
+          of
+              (AST.JoinAll, SingleLine subject') ->
+                  line $ row [ keyword "case" , space , subject' , space , keyword "of" ]
+              (_, subject') ->
+                  stack1
+                      [ line $ keyword "case"
+                      , indent subject'
+                      , line $ keyword "of"
+                      ]
+
+        (multilineBranches, branches') =
+            List.mapAccumR
+            (\mline (AST.Commented prePat pat postPat, (preRes, res)) ->
+                let
+                    prePat' =
+                        Maybe.maybeToList $ formatComments prePat
+                    pat' =
+                        formatPattern elmVersion False pat
+                            |> negativeCasePatternWorkaround pat
+                    postPat' =
+                        Maybe.maybeToList $ formatComments postPat
+                    preRes' =
+                        Maybe.maybeToList $ formatComments preRes
+                    res' =
+                        formatExpression elmVersion importInfo SyntaxSeparated res
+                    singles =
+                        allSingles $ concat [ prePat', [pat'], postPat', [line (keyword "->")], preRes', [res']]
+                in
+                case (multilineFirstBranch, singles, res') of
+                    (AST.JoinAll, Right _, SingleLine _) ->
+                        (mline, (singles, prePat', pat', postPat', preRes', res'))
+                    _ ->
+                        (AST.SplitAll, (singles, prePat', pat', postPat', preRes', res'))
+            )
+            multilineFirstBranch
+            branches
+
+        branch multiline' (singles, prePat, pat, postPat, preRes, res) =
+            case (multiline', singles, prePat, pat, postPat) of
+                (AST.JoinAll, Right singleLines, _, _, _) ->
```

Comment:
I see my previous code was similarly confusing, but it's not obvious here that `singleLines` includes the data from `pat, postPat, preRes, res`.  Could you at least add a comment right above this `case` noting that?  (without it, I started reading these branches and though "oh no, is some of the AST getting lost here?")

Reply:
I have reworked this part and it is hopefully a lot less confusing.
I used an `Either` to differentiate between single lines and boxes to avoid the variables overlap.